### PR TITLE
Keep the filter when marking a notification read or unread 

### LINKED ISSUE #9908: Marking a notification as read/unread drops filters
Peek 2020-07-13 12-08

### PR DESCRIPTION
**Pull request by danidoni**
Instead of picking the notification type from the `params` variable,
this PR makes the notification type explicitly available as a partial
local variable and function parameters.

Before this PR, the user was losing the selected notifiable type from the
notifications filter when marking a notification as 'read'.

This was due to the notifications/update remote partial not getting
the `params[:type]` when rendering.

Peek 2020-08-07 13-02

It's also fixed in small devices

Peek 2020-08-06 16-13

Fixes #9908

### CODE REVIEW

File: src/api/spec/features/webui/users/notifications_controller_spec.rb [hunk 1]
```diff
@@ -0,0 +1,30 @@
+require 'browser_helper'
+
+RSpec.describe 'User notifications', type: :feature, js: true, beta: true do
```
Comment (saraycp) [inline src/api/spec/features/webui/users/notifications_controller_spec.rb:3]: _Beta_ is set for this whole test, so I'd suggest to move it to `spec/features/beta`. All the tests in that directory has beta enabled by default, so there is no need to add `beta: true`.

That behavior comes from: 
https://github.com/openSUSE/open-build-service/blob/master/src/api/spec/support/beta.rb

**Comment by danidoni** [comment]
> Testing in review-app, it seems we have the same problem when filtering by project name.
> 
> Could that be solved in a similar way?

I broke something with the refactor, I'll take a look

**Comment by danidoni** [comment]
> > Testing in review-app, it seems we have the same problem when filtering by project name.
> > Could that be solved in a similar way?
> 
> I broke something with the refactor, I'll take a look

Fixed

File: src/api/app/models/notifications_filter.rb [hunk 2]
```diff
@@ -0,0 +1,11 @@
+class NotificationsFilter
```
Comment (danidoni) [inline src/api/app/models/notifications_filter.rb:1]: I'm not sure where to put this model

File: src/api/spec/features/beta/webui/users/notifications_controller_spec.rb [hunk 3]
```diff
@@ -0,0 +1,67 @@
+require 'browser_helper'
+
+RSpec.describe 'User notifications', type: :feature, js: true, beta: true do
```
Comment (eduardoj) [inline src/api/spec/features/beta/webui/users/notifications_controller_spec.rb:3]: As @saraycp pointed out, after moving the test into `spec/features/beta/`, you don't need to define the `beta: true` metadata.

File: src/api/app/models/notifications_filter.rb (hunk 2 above)
Comment (dmarcoux) [reply src/api/app/models/notifications_filter.rb:1]: This looks like a presenter. A quite basic one, but still a presenter.

File: src/api/spec/features/beta/webui/users/notifications_controller_spec.rb (hunk 3 above)
Comment (danidoni) [reply src/api/spec/features/beta/webui/users/notifications_controller_spec.rb:3]: True, I'll remove it

File: src/api/app/models/notifications_filter.rb (hunk 2 above)
Comment (danidoni) [reply src/api/app/models/notifications_filter.rb:1]: Hmm, yes I thought about it before, but I wasn't sure. I'll try again.

File: src/api/spec/features/beta/webui/users/notifications_controller_spec.rb (hunk 3 above)
Comment (danidoni) [reply src/api/spec/features/beta/webui/users/notifications_controller_spec.rb]: Done

File: src/api/app/models/notifications_filter.rb (hunk 2 above)
Comment (danidoni) [reply src/api/app/models/notifications_filter.rb]: I've renamed the class as a presenter

File: src/api/spec/features/beta/webui/users/notifications_controller_spec.rb [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    let(:project) { notifiable.commentable.project }
+
+    context 'when clicking on the Comments filter' do
+      before do
+        login user
+        visit my_notifications_path
+      end
+
+      it 'shows all unread comment notifications' do
+        find(class: 'fa-chevron-down').click if mobile?
```
Comment (dmarcoux) [inline src/api/spec/features/beta/webui/users/notifications_controller_spec.rb:33]: This seems quite fragile. Could we have a selector which is more specific, something like `#notifications-filter-desktop i.fa-chevron-down`?

Comment (danidoni) [reply src/api/spec/features/beta/webui/users/notifications_controller_spec.rb:33]: Yes, good idea, I did that before, no reason to not doing it now.

Comment (danidoni) [reply src/api/spec/features/beta/webui/users/notifications_controller_spec.rb:33]: Even better, I'll add an id to the dropdown control

File: src/api/spec/features/beta/webui/users/notifications_controller_spec.rb [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    let(:project) { notifiable.commentable.project }
+
+    context 'when clicking on the Comments filter' do
+      before do
+        login user
+        visit my_notifications_path
+      end
+
+      it 'shows all unread comment notifications' do
+        find(id: '#notifications-dropdown-trigger').click if mobile?
```
Comment (dmarcoux) [inline src/api/spec/features/beta/webui/users/notifications_controller_spec.rb:33]: To be consistent with other feature specs: 
```suggestion
        find('#notifications-dropdown-trigger').click if mobile?
```

File: src/api/spec/features/beta/webui/users/notifications_controller_spec.rb [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+        within('#filters') { click_link('Comments') }
+        expect(page).to have_text(notifiable.commentable.name)
+      end
+
+      context 'when marking a comment notification as read' do
+        before do
+          find(id: '#notifications-dropdown-trigger').click if mobile?
+          within('#filters') { click_link('Comments') }
+          find_link(id: format('update-notification-%d', notification_for_projects_comment.id)).click
```
Comment (dmarcoux) [inline src/api/spec/features/beta/webui/users/notifications_controller_spec.rb:42]: This should work, right?
```suggestion
          click_link("update-notification-#{notification_for_projects_comment.id}")
```

Comment (danidoni) [reply src/api/spec/features/beta/webui/users/notifications_controller_spec.rb:42]: Yyyesss

File: src/api/spec/features/beta/webui/users/notifications_controller_spec.rb [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+        visit my_notifications_path
+      end
+
+      it 'shows all unread comment notifications' do
+        find('#notifications-dropdown-trigger').click if mobile?
+        within('#filters') { click_link('Comments') }
+        expect(page).to have_text(notifiable.commentable.name)
+      end
+
+      context 'when marking a comment notification as read' do # rubocop:todo RSpec/NestedGroups
```
Comment (danidoni) [inline src/api/spec/features/beta/webui/users/notifications_controller_spec.rb:38]: What about disabling this at project level? I find it too tight

Comment (dmarcoux) [reply src/api/spec/features/beta/webui/users/notifications_controller_spec.rb:38]: I would be fine with setting the maximum level of nesting to `2`.

https://rubocop.readthedocs.io/projects/rspec/en/latest/cops_rspec/#rspecnestedgroups

Comment (danidoni) [reply src/api/spec/features/beta/webui/users/notifications_controller_spec.rb:38]: Then let's do it

File: .rubocop.yml [hunk 8]
```diff
@@ -31,6 +32,11 @@ Layout/DotPosition:
 Layout/EndAlignment:
   EnforcedStyleAlignWith: variable
 
+#################### RSpec ###########################
+
+RSpec/NestedGroups:
```
Comment (dmarcoux) [inline .rubocop.yml:37]: I don't mind having this in the PR, but it should be in a different commit. It's not related to the notification filters.

Comment (danidoni) [reply .rubocop.yml:37]: Yes, I'll do it in another PR, because including RSpec cops is uncovering a lot of issues. I'm going back to having it disabled locally from that specific spec.

File: src/api/app/views/webui/users/notifications/index.html.haml [hunk 9]
```diff
@@ -9,8 +9,9 @@
...
                                   aria: { expanded: true, controls: 'filters' } }
         Filtered by: #{params[:type]&.humanize || @filtered_project || 'Unread'}
-        %i.float-right.mt-1.fa.fa-chevron-down
+        %i.float-right.mt-1.fa.fa-chevron-down#notifications-dropdown-trigger
       .card-body.collapse#filters
-        = render partial: 'notifications_filter', locals: { projects_for_filter: @projects_for_filter, notifications_count: @notifications_count }
+        = render partial: 'notifications_filter', locals: { filter: @notifications_filter }
   .col-md-8.col-lg-9#notifications-list
-    = render partial: 'notifications_list', locals: { notifications: @notifications }
+    - notifications_list_locals = { notifications: @notifications, selected_filter: @notifications_filter.selected_filter }
```
Comment (krauselukas) [inline src/api/app/views/webui/users/notifications/index.html.haml:16]: @danidoni its maybe nitpicking, but I would skip the step of assigning this to a variable, and just pass it directly to the locals

Comment (danidoni) [reply src/api/app/views/webui/users/notifications/index.html.haml]: Yes, fixed

**Comment by krauselukas** [comment]
@danidoni LGTM, just squash the last commit and I will approve
